src/toolkits/activity_classification/ac_data_iterator.cpp

```diff
@@ -132,152 +132,291 @@ variant_map_type _activity_classifier_prepare_data_impl(const gl_sframe &data,
+  number_of_sessions++;
 
-    size_t number_of_sessions = 0;
+  if (verbose) {
+    logprogress_stream << "Processed a total of " << number_of_sessions
+                       << " sessions." << std::endl;
+  }
+  gl_sframe converted_sframe = output_writer.close();
+  converted_sframe.materialize();
 
-    // Prepare an output SFrame writer, that will write a new SFrame in the converted batch-processing
-    // ready format.
-    std::vector<std::string> output_column_names = {"features", "chunk_len", "session_id"};
-    std::vector<flex_type_enum> output_column_types = {flex_type_enum::VECTOR, flex_type_enum::INTEGER,
-                                                        data[session_id].dtype()};
-    if (use_target) {
-        curr_chunk_targets.reserve(predictions_in_chunk);
-        curr_window_targets.reserve(prediction_window);
-        output_column_names.insert(output_column_names.end(), {"target", "weights"});
-        output_column_types.insert(output_column_types.end(), {flex_type_enum::VECTOR, flex_type_enum::VECTOR});
-    }
-    gl_sframe_writer output_writer(output_column_names, output_column_types, 1);
+  variant_map_type result_dict;
+  result_dict["converted_data"] = converted_sframe;
+  result_dict["num_of_sessions"] = number_of_sessions;
 
-    if (verbose) {
-      logprogress_stream << "Using sequences of size " << chunk_size << " for model creation." << std::endl;
-    }
+  return result_dict;
+}
 
-    time_t last_print_time = time(0);
-    size_t processed_lines = 0;
-
-    // Iterate over the user data. The features and targets are aggregated, and handled
-    // whenever a the ending of a prediction_window, chunk or session is reached.
-    for (const auto& line: data.range_iterator()) {
-
-        auto curr_session_id = line[column_index_map[session_id]];
-
-        // Check if a new session has started
-        if (curr_session_id != last_session_id) {
-
-            // Finalize the last chunk of the previous session
-            if (curr_chunk_features.size() > 0) {
-                finalize_chunk(curr_chunk_features,
-                               curr_chunk_targets,
-                               curr_window_targets,
-                               last_session_id,
-                               output_writer,
-                               chunk_size,
-                               feature_size,
-                               predictions_in_chunk,
-                               use_target);
-            }
-
-            last_session_id = curr_session_id;
-            number_of_sessions++;
-        }
+variant_map_type _activity_classifier_prepare_data_aug_impl(const gl_sframe &data,
+                                                        const std::vector<std::string> &features,
+                                                        const std::string &session_id,
+                                                        const int &prediction_window,
```

Comment:
There’s little reason to pass integers by const reference. May as well pass by value and have simpler, more readable code

Reply:
This is what we do in ```_activity_classifier_prepare_data_impl``` also.

Reply:
You shouldn't take that function as an ideal example of readable C++ code.

But for now, let's leave it alone, since we expect to deprecate/delete it later this year. Then we don't have to worry about additional testing that the Python side still works after your changes.